blue_bringup/scripts/gripper_calibration_service.py

```diff
@@ -0,0 +1,149 @@
+            return motor_state_msg.position[i]
+
+def handle_load_controllers():
+    unload_list = []
+    load_controllers_list = ['blue_controllers/gripper_torque_controller', 'blue_controllers/gripper_controller']
+    loaded_controllers    = list_controllers()
+    rospy.logerr(loaded_controllers.controller)
+    for controller in load_controllers_list:
+        load = True
+        for loaded in loaded_controllers.controller:
+            rospy.logerr(loaded.name)
+            rospy.logerr(controller)
+            rospy.logerr(controller)
+
+            if loaded.name == controller:
+                # controller already loaded
+                if loaded.state == "running":
+                    return False
+                load = False
+                break
+        if load:
+            # controller not loaded yet
+            controller_response = load_controller(controller)
+            unload_list.append(controller)
+    return unload_list
+
+
+def handle_calibration_service(input_request):
+    # Calibrate joints with startup angles
+    rospy.loginfo("Starting calibration...")
+    try:
+        unload_list = handle_load_controllers()
+        if unload_list == False:
+            return TriggerResponse(False, "Gripper Calibration Failed. Gripper resource must be free.")
+        gripper_controller_response = switch_controller(['blue_controllers/gripper_torque_controller'], [], 2)
+
+        # Close the gripper
+        cmd = Float64MultiArray()
+        cmd.data = [5.0]
+        for i in range(5):
+            cmd_pub.publish(cmd)
+            rospy.sleep(0.1)
+
+        # Wait until gripper is closed/stops moving
+        gripper_state = get_gripper_state(motor_names)
+        current_position = gripper_state + 10
+        while not rospy.is_shutdown() and np.abs(current_position - gripper_state) > 1e-5:
+            cmd_pub.publish(cmd)
+            current_position = gripper_state
+            gripper_state = get_gripper_state(motor_names)
+
+        # Stop applying force
+        cmd.data = [0.0]
+        for i in range(5):
+            cmd_pub.publish(cmd)
+            rospy.sleep(0.1)
+
+        # Calibrate gripper to closed position
+        # TODO: don't hardcode this value -- read from yaml? URDF?
+        response = gripper_position_calibration(1.0)
+
+        if response.success:
+            rospy.loginfo("Gripper position calibration succeeded!")
+        else:
+            rospy.logerr("Gripper position calibration failed!")
+
+        # Switch out controllers to reset position
+        gripper_controller_response = switch_controller(['blue_controllers/gripper_controller'], ['blue_controllers/gripper_torque_controller'], 2)
+
+        client = actionlib.SimpleActionClient(
+            "blue_controllers/gripper_controller/gripper_cmd",
+            GripperCommandAction,
+        )
+
+        # Wait 10 Seconds for the gripper action server to start or exit
+        if not client.wait_for_server(rospy.Duration(10.0)):
```

Comment:
just wondering -- is there a reason we actually need to re-open the gripper? anything wrong with just leaving it closed? alternatively, could we just use the torque controller to open it to simplify this script?

Reply:
No specific reason why we are reopening the gripper. I thought it could be nicer to have the gripper open. Mechanically it could be bad to open the gripper with the torque controller as the nut could get stuck. I'm ok with just keeping the gripper closed though and it would simplify the code.

Reply:
I went ahead and removed the opening of the gripper after calibration.

Reply:
Ok! We also have position feedback, so if we did want to use the torque controller to re-open the gripper we could probably use that to prevent it from getting stuck.